### PR TITLE
ASoC: SOF: pm: save io region state in case of errors in resume 

### PR DESCRIPTION
**Pull request by kv2019i**
ASoC: SOF: pm: save io region state in case of errors in resume

If there are failures in DSP runtime resume, the device state will not
reach active and this makes it impossible e.g. to retrieve a possible
DSP panic dump via "exception" debugfs node. If
CONFIG_SND_SOC_SOF_DEBUG_ENABLE_DEBUGFS_CACHE=y is set, the data in
cache is stale. If debugfs cache is not used, the region simply cannot
be read.

To allow debugging these scenarios, update the debugfs cache contents in
resume error handler. User-space can then later retrieve DSP panic and
other state via debugfs (requires SOF debugfs cache to be enabled in
build).

Reported-by: Curtis Malainey <cujomalainey@chromium.org>
Link: https://github.com/thesofproject/linux/issues/4274
Signed-off-by: Kai Vehmanen <kai.vehmanen@linux.intel.com>

### CODE REVIEW

**Comment by kv2019i** [comment]
Tested with error injection using following two patches:
- kernel change to send special IPC message on resume -> https://github.com/kv2019i/linux/commit/ef14872c363f34fdcc249e9736feebef74987c00
- FW change (on top of stable-v2.2) to panic in case special IPC message is received -> https://github.com/kv2019i/sof/commit/902a684567da741f4399c536dc68e805e159cae5

File: sound/soc/sof/pm.c [hunk 1]
```diff
@@ -178,6 +178,18 @@ static int sof_resume(struct device *dev, bool runtime_resume)
 			dev_err(sdev->dev, "ctx_restore IPC error during resume: %d\n", ret);
 	}
 
+setup_fail:
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_ENABLE_DEBUGFS_CACHE)
+	if (ret < 0) {
+		/*
+		 * Debugfs cannot be read in runtime suspend, so cache
```
Comment (plbossart) [inline sound/soc/sof/pm.c:185]: I read that comment multiple times, and I don't think it's correct.

We *can* read debugs, it's just that in pm_runtime suspend the cache contents will be provided.

This comment confuses cause and effect IMHO.

the commit message has the same issue, it's not "impossible to retrieve an exception", it's rather "the exception is not available in the cached data".

I may be splitting hair on this one but the logic seems off.

Comment (cujomalainey) [reply sound/soc/sof/pm.c:185]: @plbossart the issue is I think we cannot force DSP to remain in D0 if it fails to enter D0, which means we cannot collect core dumps because the DSP is left in D3. This caches them on failure so that we can still get them even though DSP never entered D0.

File: sound/soc/sof/pm.c [hunk 2]
```diff
@@ -164,7 +164,7 @@ static int sof_resume(struct device *dev, bool runtime_resume)
 		ret = tplg_ops->set_up_all_pipelines(sdev, false);
```
Comment (cujomalainey) [inline sound/soc/sof/pm.c:164]: nit for metrics on our side: Reported-by: Curtis Malainey <cujomalainey@chromium.org>

File: sound/soc/sof/pm.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/pm.c:185]: I get the issue, what I argue on is the comments and explanations. We can *always* read debugfs, it's only the cached results that need to be forced.

Comment (cujomalainey) [reply sound/soc/sof/pm.c:185]: Regarding your comment though, in the linked failure state, it is impossible to fetch the failure state, with or without cache without this change.

With cache you get nothing
Without cache you get told DSP is in wrong state and get nothing (i.e cannot be read)

Comment (kv2019i) [reply sound/soc/sof/pm.c:185]: Thanks, let me rephrase the commit message and reupload.

File: sound/soc/sof/pm.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/pm.c:164]: missed Curtis' Reported-by in the update

**Comment by kv2019i** [comment]
V2 uploaded:
- rewrote both the git commit description and the inline comment
- added reported-by tag

**Review by plbossart — APPROVED** [review]

**Review by ranj063 — APPROVED** [review]

**Review by cujomalainey — APPROVED** [review]

**Review by ujfalusi — APPROVED** [review]
